### DISCUSSION TITLE
link-layer - general

### DISCUSSION THREAD

**Discussion by ivanhofer** [pipeli18ne/RFC#4] [Feedback]
Comments about the proposed solution to the link-layer problem

**Comment by samuelstroschein** [comment]
> Summary: a single library can't offer all points a team will consider when choosing a i18n library to use in their project. You will have to make compromises. 

Strongly agree with this summary. We, referring to inlang, made the mistake of focusing on one library. The edge cases are too vast. One ready-made library without a plugin system or similar is unlikely to succeed. 

Quoting from inlang's RFC 001:

> #### Decision
> Leverage existing i18n libraries instead of forcing a specific inlang library.
> 
> [...] One library to rule them all is unlikely a feasible idea. Applications have different requirements.

## Overall, the link-layer proposal seems to be a step in the right direction

I am biased though. (We discussed this privately) By chance, we simultaneously concluded that a `link-layer` is of benefit to ease localization through customization and plugins _and started to draft solutions without being aware of the counterpart solution_. That we, in my eyes, arrived at the same conclusion is a testament that the proposals "are more correct than incorrect". We just choose different words for the building blocks. For example, we call the `link-layer` the `inlang.config.js` file. 

```mermaid
flowchart BT

    subgraph Components
      Library
      DT[Dev Tools]
      CICD[CI/CD]
      Editor
    end

    Config <--> |read & write callbacks| Resources
    Config <--> |uses config functions as business logic| Components
```
_Graph is taken from inlang RFC-001 ["How is inlang configured?"](https://github.com/inlang/inlang/edit/main/rfcs/001-core-architecture.md#2-how-is-inlang-configured) and illustrates the `link-layer` of the config. Resources are read from files, components (called building blocks in pipel18ne's RFC) can leverage the config to share business logic._

Since we use different terms, looking at our terms might be of benefit. We leaned heavily towards industry standard terms from Fluent or the new MessageFormat 2.0. 
 
> `dictionaries`: the files where your translated strings are located

We use `resources` which aligns with Fluent and MessageFormat 2.0. A resource also does not have to be a dictionary data structures per-se. A CSV could serve as "resource" as well.   

- `metadata-blocks`: an object representation of translations  

We use the term AST which you likely also mean with object representation of translations. Talking about "translations", "translations" are usually referred to as `Messages`. The base language (we use reference language) "messages" are no translations. Technically a translation is nothing else than a Message in a different variant.

**Reply by ivanhofer** [reply]
Thanks for your thoughts!
I think the terms used in this RFC are "more common" to the terminology of most i18n solutions. Since this is just a first draft, terms will probably change once things get more concrete.

> We use the term AST which you likely also mean with object representation of translations

I also have thought of using `AST`, but I thought `metadata-representation` would be the better word, because this not only represents source code, but also metadata on top. e.g. in what file is this `translation` (or `message` 😉) used or other manually defined meta information.
